pkg/cgroups/cgroupblkio.go

```diff
@@ -106,6 +109,143 @@ func GetBlkioDir() string {
+	seenDev := map[devMM]bool{}
+	for _, rdp := range wanted {
+		new = append(new, rdp)
+		seenDev[devMM{rdp.Major, rdp.Minor}] = true
+	}
+	for _, rdp := range old {
+		if !seenDev[devMM{rdp.Major, rdp.Minor}] {
+			new = append(new, OciRateDeviceParameters{rdp.Major, rdp.Minor, 0})
+		}
+	}
+	return new
+}
+
+// GetBlkioParameters returns OCI BlockIO parameters from files in cgroups blkio controller directory.
+func GetBlkioParameters(cgroupsDir string) (OciBlockIOParameters, error) {
+	var errors *multierror.Error
+	blockIO := NewOciBlockIOParameters()
+	content, err := readFromFileInDir(cgroupsDir, blkioWeightFiles)
+	if err == nil {
+		weight, err := strconv.ParseInt(strings.TrimSuffix(content, "\n"), 10, 64)
+		if err == nil {
+			blockIO.Weight = weight
+		} else {
+			errors = multierror.Append(errors, fmt.Errorf("parsing weight from %#v failed: %w", content, err))
+		}
+	} else {
+		errors = multierror.Append(errors, err)
+	}
+	errors = multierror.Append(errors, readDevWeightsRates(cgroupsDir, blkioWeightDeviceFiles, &blockIO.WeightDevice, nil))
+	errors = multierror.Append(errors, readDevWeightsRates(cgroupsDir, blkioThrottleReadBpsFiles, nil, &blockIO.ThrottleReadBpsDevice))
+	errors = multierror.Append(errors, readDevWeightsRates(cgroupsDir, blkioThrottleWriteBpsFiles, nil, &blockIO.ThrottleWriteBpsDevice))
+	errors = multierror.Append(errors, readDevWeightsRates(cgroupsDir, blkioThrottleReadIOPSFiles, nil, &blockIO.ThrottleReadIOPSDevice))
+	errors = multierror.Append(errors, readDevWeightsRates(cgroupsDir, blkioThrottleWriteIOPSFiles, nil, &blockIO.ThrottleWriteIOPSDevice))
+	return blockIO, errors.ErrorOrNil()
+}
+
+func readDevWeightsRates(baseDir string, filenames []string, outWeights *[]OciWeightDeviceParameters, outRates *[]OciRateDeviceParameters) error {
+	mmvs, err := readMajorMinorValues(baseDir, filenames)
+	if err != nil {
+		return err
+	}
+	if outWeights != nil {
+		for _, mmv := range mmvs {
+			*outWeights = append(*outWeights, OciWeightDeviceParameters{mmv.major, mmv.minor, mmv.value})
+		}
+	}
+	if outRates != nil {
+		for _, mmv := range mmvs {
+			*outRates = append(*outRates, OciRateDeviceParameters{mmv.major, mmv.minor, mmv.value})
+		}
+	}
+	return nil
+}
+
+var reMajorMinorValue = regexp.MustCompile("([0-9]+):([0-9]+) ([0-9]+)")
+
+type majorMinorValue struct {
+	major int64
+	minor int64
+	value int64
+}
+
+func readMajorMinorValues(baseDir string, filenames []string) ([]majorMinorValue, error) {
+	mmvs := []majorMinorValue{}
+	content, err := readFromFileInDir(baseDir, filenames)
+	if err != nil {
+		return mmvs, err
+	}
+	for _, submatches := range reMajorMinorValue.FindAllStringSubmatch(content, -1) {
+		if len(submatches) == 4 {
+			major, _ := strconv.ParseInt(submatches[1], 10, 64)
+			minor, _ := strconv.ParseInt(submatches[2], 10, 64)
+			value, _ := strconv.ParseInt(submatches[3], 10, 64)
+			mmvs = append(mmvs, majorMinorValue{
+				major: major,
+				minor: minor,
+				value: value,
+			})
+		}
+	}
```

Comment:
Maybe it's just me, but, I dislike regexes in simple cases like these. SImple strings split should be good, and, I'd print an error/warning if we encounter unparsable input.

WDYT?

Reply:
Ok, rewrote the parser without regexps and with error messages on unexpected lines.